### PR TITLE
Support api v3

### PR DESCRIPTION
**Pull request by fantostisch**
Adds support for [API V3](https://github.com/eduvpn/documentation/blob/v3/API.md) including WireGuard.

This pull requests includes https://github.com/eduvpn/android/pull/347, https://github.com/eduvpn/android/pull/344, https://github.com/eduvpn/android/pull/343 and https://github.com/eduvpn/android/pull/333.

#### Other notable changes

The OpenVPN notification has been replaced with our own notification:
image
The WireGuard library does not provide a notification, which means we have to create our own, this notification is now also used for OpenVPN.

Most JSON (de)serialization code has been removed and is now generated at compile time using [Kotlin serialization](https://github.com/Kotlin/kotlinx.serialization), which means we do not have to write trivial JSON serialization code anymore. This plugin is officially supported by JetBrains so should be safe to use.

The connection log only shows logs for OpenVPN. Reading the WireGuard logs is possible, but not trivial as it would require starting a logcat process and than parsing its output. A possible solution would be to copy-paste [the relevant code](https://git.zx2c4.com/wireguard-android/tree/ui/src/main/java/com/wireguard/android/activity/LogViewerActivity.kt?h=1.0.20211029&id=1d7d306950082040c41f3b627d57e8c802b4b792#n176) from the WireGuard Android app. I tried to add the WireGuard Android app as a submodule, but I was unable to integrate their gradle setup.

Profile names now support translations. The server does not support this (yet?) but it is specified in the API.

Connecting over an insecure connection (HTTP) was already disabled automatically by Android on Android >=9, but is now also manually disabled on Android >= 6. Redirects from HTTPS to HTTP are now disabled on Android 5 as specified by the API spec.

#### Other notes

There is currently no easy way for the user to know if the app is using OpenVPN or WireGuard. You could check if the connection log is empty or If you have access to the server you can check if the assigned IP is in the WireGuard or OpenVPN range. Another possibility is looking at the logcat. Do we want to show the used VPN protocol?

No online detection is implemented. Who do we connect to?

When checking out this branch, don't forget to also update the submodules which can be done using `git submodule deinit -f . && git submodule update --init --recursive`.

### CODE REVIEW

**Comment by fantostisch** [comment]
To make reviewing a bit easier, I made a diff (https://github.com/fantostisch/android/commit/5bc5bfd50cf22ac9181e6e80987fee9c4929a6bf) that does not include the other pull requests and does not include the conversions to Kotlin. The excluded commits can be seen at https://github.com/eduvpn/android/compare/master...fantostisch:diff4

File: app/src/main/java/nl/eduvpn/app/service/WireGuardService.kt [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+    private val tunnel = WireGuardTunnel("eduVPN WireGuard tunnel") { newTunnelState ->
+        setConnectionStatus(tunnelStateToStatus(newTunnelState))
+    }
+
+    override fun getStatus(): VPNStatus {
+        return connectionStatus
+    }
+
+    override fun startForeground(id: Int, notification: Notification) {
+        // Not necessary for WireGuard
```
Comment (dzolnai) [inline app/src/main/java/nl/eduvpn/app/service/WireGuardService.kt:45]: Does WireGuard have its own notification? Or why don't we show it here?

File: app/src/main/java/nl/eduvpn/app/service/WireGuardService.kt [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+                } else {
+                    fail(ex.toString())
+                }
+            }
+        }
+    }
+
+    private fun fail(errorString: String) {
+        setConnectionStatus(VPNStatus.FAILED)
+        this.errorString = errorString
```
Comment (dzolnai) [inline app/src/main/java/nl/eduvpn/app/service/WireGuardService.kt:103]: You should set the error string before notifying the subscribers about the failure

File: app/src/main/java/nl/eduvpn/app/utils/GeneralExtensions.kt [hunk 3]
```diff
@@ -121,3 +122,17 @@ inline fun <R> runCatchingCoroutine(block: () -> R): Result<R> {
     }
     return result
 }
+
+val jsonInstance: Json by lazy {
```
Comment (dzolnai) [inline app/src/main/java/nl/eduvpn/app/utils/GeneralExtensions.kt:126]: I think this can be better put inside SerializerService

**Review by dzolnai — CHANGES_REQUESTED** [review]

**Comment by dzolnai** [comment]
I tested the application on this branch, and it seems to work fine, including the WireGuard protocol implementation. 👍

File: app/src/main/java/nl/eduvpn/app/MainActivity.kt [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+        private val TAG = MainActivity::class.java.name
+        private const val REQUEST_CODE_SETTINGS = 1001
+        private const val KEY_BACK_NAVIGATION_ENABLED = "back_navigation_enabled"
+    }
+
+    @Inject
+    protected lateinit var historyService: HistoryService
+
+    @Inject
+    protected lateinit var vpnService: Optional<VPNService>
```
[the remark this answers is not in the document]
Comment (fantostisch) [reply app/src/main/java/nl/eduvpn/app/MainActivity.kt:62]: The MainActivity needs a vpnService, but when opening the app for the first time no vpn connection has been made yet, so we do not know if we need the WireGuard or the OpenVPN service. The empty case is used to indicate that no vpn protocol has been chosen yet.

File: app/src/main/java/nl/eduvpn/app/entity/DiscoveredAPIs.kt [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+data class DiscoveredAPIs(
+
+    @SerialName("http://eduvpn.org/api#2")
+    val v2: DiscoveredAPIV2? = null,
+
+    @SerialName("http://eduvpn.org/api#3")
+    val v3: DiscoveredAPIV3? = null,
+) {
+    fun getPreferredAPI(): DiscoveredAPI? {
+        return v2 ?: v3
```
[the remark this answers is not in the document]
Comment (fantostisch) [reply app/src/main/java/nl/eduvpn/app/entity/DiscoveredAPIs.kt]: I changed it to prefer v3, but I think it does not really matter. When we remove v2 support from the Android app, everyone is forced to use v3 API anyway.

File: app/src/main/java/nl/eduvpn/app/entity/exception/EduVPNException.kt [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+
+/**
+ * Exception with translatable message.
+ *
+ * @param formatArgs Arguments for [resourceIdMessage]
+ */
+class EduVPNException(
+    @StringRes val resourceIdTitle: Int,
+    @StringRes val resourceIdMessage: Int,
+    vararg formatArgs: Any?
```
[the remark this answers is not in the document]
Comment (fantostisch) [reply app/src/main/java/nl/eduvpn/app/entity/exception/EduVPNException.kt:13]: > Can't you mark this a val, so you don't need line 15?

there seems to be no way to remove line 15, "val vararg" or other variations do not seem to be possible in Kotlin

> I would add a default null value for this param, not all messages need format args.

in the case of a message without parameters, you can just omit the varargs, no null necessary, see for example https://github.com/eduvpn/android/blob/a28efb42742732c0f8f096c50b83e3f21d30bd5e/app/src/main/java/nl/eduvpn/app/service/ConnectionService.kt#L198

File: app/src/main/java/nl/eduvpn/app/service/PreferencesService.kt [hunk 7]
```diff
@@ -0,0 +1,723 @@
...
+ */
+class PreferencesService(
+    applicationContext: Context,
+    serializerService: SerializerService,
+    securePreferences: SharedPreferences
+) {
+    companion object {
+        private val TAG = PreferencesService::class.simpleName
+
+        private const val STORAGE_VERSION = 3
```
[the remark this answers is not in the document]
Comment (fantostisch) [reply app/src/main/java/nl/eduvpn/app/service/PreferencesService.kt:50]: The master branch shows version 3: https://github.com/eduvpn/android/blob/8fa7c817a3b556d341efdb266940658ac8b32ec1/app/src/main/java/nl/eduvpn/app/service/PreferencesService.java#L55

Somewhere in the git history I did create version 4 and then removed it as it was not necessary.

I do not plan to do a clean start. No upgrade path is necessary, the serialization code was modified, but the data format was not changed. When we remove v2 support can remove the v2 specific data.

**Comment by fantostisch** [comment]
All feedback/issues without a reply have been fixed, see commits above ^

File: app/src/androidTest/java/nl/eduvpn/app/service/PreferencesServiceTest.kt [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+
+/**
+ * Tests for the preferences service.
+ * Created by Daniel Zolnai on 2016-10-18.
+ */
+@RunWith(AndroidJUnit4::class)
+@LargeTest
+class PreferencesServiceTest {
+
+    private var _preferencesService: PreferencesService? = null
```
Comment (dzolnai) [inline app/src/androidTest/java/nl/eduvpn/app/service/PreferencesServiceTest.kt:40]: You could make these a lateinit var and then you can avoid the force casts

**Review by dzolnai — APPROVED** [review]

File: app/src/androidTest/java/nl/eduvpn/app/service/PreferencesServiceTest.kt (hunk 8 above)
Comment (dzolnai) [reply app/src/androidTest/java/nl/eduvpn/app/service/PreferencesServiceTest.kt]: Pushed my fix for this on the branch so we can merge

**Pull request merged**